### PR TITLE
perf: make escapeForJson cost linear-ish and content-independent (#42)

### LINKED ISSUE #42: escapeForJson is quadratic, and its comment claims O(n)
`escapeForJson`'s header comment says:

> Use text item delimiters for O(n) performance instead of character-by-character O(n²)

That is true of the five delimiter phases. It is not true of the handler, because the per-character loop beneath them dominates — and the loop is quadratic. Measured on an M-series Mac, escaping a string of plain `a`:

| input | time |
|---|---|
| 1,000 chars | 0.09s |
| 5,000 chars | 0.74s |
| 20,000 chars | 10.50s |

4× the input for 14× the time. `osascript` startup is ~0.03s, so that is the loop, not process overhead.

**Why it matters.** `runAppleScript` uses `DEFAULT_TIMEOUT` of 30s. Extrapolating the curve, a body around 40,000 characters reaches it — a long email thread with quoted history is not an unusual size. The failure would surface as a timeout, which names neither the tool nor the cause.

**Why it is not urgent.** The read path escapes subjects and senders, which are short. The compose path escapes bodies, which are agent-supplied and in practice have been small. Nothing has hit this.

**Two things to separate when fixing.** The claim and the behaviour are independent defects, and DoD 4 covers the first on its own: the comment should stop asserting O(n) for the handler whatever happens to the loop. That part is a one-line correction and could ship immediately.

For the loop itself, the likely cause is `copy … to end of resultList` — AppleScript list append is not amortised constant, so this is the classic accumulate-in-a-list shape. Worth measuring a preallocated-list or chunked-string approach before assuming which is faster.

Measured while addressing the supervisor round on #41; recorded rather than fixed there, since changing the loop's shape in a correctness fix would have obscured both.

### PR DESCRIPTION
**Pull request by marius-cetanas**
Closes #42.

## What was wrong

Escaping 40,000 characters took **~25s** against the bridge's 30s `DEFAULT_TIMEOUT`. A body somewhat
larger than that failed as a timeout naming neither the tool nor the cause. The same input now
escapes in **~0.3s**.

Measured on an M-series Mac, realistic email prose, wall clock including ~0.05s of `osascript`
startup:

| input | before | after |
|---|---|---|
| 40,000 | 27.67s | 0.28s |
| 100,000 | — | 0.74s |
| 200,000 | — | 2.45s |

## The diagnosis in the issue was half right, and that is why this took measuring

#42 guessed the cause was `copy … to end of`. That **is** a real quadratic source, and it was not
the only one — element access was quadratic too, and **each masked the other**. Replacing `copy … to
end of` with `set end of` alone measures as *no improvement at all* (5.80s vs 5.82s at 20,000),
which is exactly the reading that would have closed this as "not the cause".

Four list operations were measured to be traps, and all four have to be avoided together:

| trap | measured | what replaces it |
|---|---|---|
| `item i of aList` on a plain variable | read loop alone, no escaping, no appends: 0.08s at 50,000 → **86.45s** at 100,000 | the list lives in a script object property |
| `items a thru b of aBigList` | costs in proportion to `a`, so slicing a run per escape is quadratic in escape count | runs accumulate in a small bounded buffer |
| `count of aList` in the loop body | not free; a variant using it cost 20.44s where the same shape with integer counters cost 0.28s | lengths tracked in integers |
| unbounded appends | grows the target list until it re-enters trap 1 | both accumulators flush at a fixed threshold |

## Content-independence is the other half of the fix

Before, cost depended on *how much* of the input needed escaping — the same 40,000 characters cost
an order of magnitude more as quotes than as prose. After, it is flat:

| escape density | 0% | 5% | 10% | 20% | 50% |
|---|---|---|---|---|---|
| previous best candidate | 0.18s | 0.23s | 0.33s | 0.77s | 3.06s |
| **this** | 0.33s | 0.25s | 0.25s | 0.27s | 0.29s |

## Correctness is untouched

Every code point is still judged once, on its own, from the flat `id of` list — the invariant #41
established. All **30** existing executed assertions pass unchanged, including every #33 / #39 / #41
grapheme-cluster case.

## The new assertions were verified by mutation

Three timing assertions ship with the claim (DoD 3). Against the previous handler they **fail at
24.9s over a 10s bound** — checked, not assumed. One of them pins content-independence, which no
single-input timing test can.

The bound is not a target: it is ~30× the measured cost, there to catch a return to the quadratic
shape rather than to police drift on a slower machine. Like every executed AppleScript test here,
these run on **macOS only and never in the merge gate**.

## What this does not claim

It is **not O(n)**, and the header comment does not say it is — past roughly 65,000 code points the
one unavoidable walk degrades again. A false complexity claim in that very header was the other half
of what #42 reported, so replacing it with a different false claim seemed worth avoiding.

## Note for reviewers

`escape-for-json.applescript` now defines **two** handlers, and the bridge prepends both to every
script — so `joinStrings` is now a reserved name in domain scripts. `CLAUDE.md` records that.

## Verify

`npm run build && npm run test:coverage && npm audit --audit-level=high` — green locally on macOS:
**494 tests across 27 files**, 100% statements/branches/functions/lines on `src/`, 0 vulnerabilities.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/helpers/run-applescript.ts [hunk 1]
```diff
@@ -14,12 +14,25 @@ import { join } from "node:path";
  * `osascript` exists, so callers gate on `darwin` — see `onMacOS` below.
  */
 export function runHandler(handlerPath: string, expression: string): string {
+  return runHandlerScript(handlerPath, `return ${expression}`);
+}
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply tests/helpers/run-applescript.ts:37]: Fixed in b706453. You were right that it was a regression: extracting `runHandlerScript` left `runHandler` inheriting the 120s default, so all 30 correctness assertions — which escape a handful of code points each — would hold the suite for two minutes on a hang instead of failing in twenty seconds.

They are now two named constants rather than one shared default. They cannot be merged: a regression to the quadratic shape #42 reported takes ~25s, so putting the large-input tests on the 20s timeout would turn their assertion into an opaque `ETIMEDOUT` from the runner instead of the measured number the test exists to report.

**Comment by marius-cetanas** [comment]
## Review round addressed — including three claims of mine that were false

The rewrite itself held: 306 adversarial differential cases against the old handler, byte-identical
on every one, and the headline numbers reproduce. The prose around it did not, and two of the
corrections are to exactly the kind of claim this header exists in order **not** to make.

### A measurement artifact I wrote in as fact

The header claimed element access "falls off a cliff between 50,000 and 100,000 elements — 0.08s to
86s". **There is no cliff.** Re-measured:

| elements | read loop, plain variable |
|---|---|
| 20,000 | 3.36s |
| 35,000 | 10.16s |
| 50,000 | 20.46s |
| 65,000 | 35.01s |

Smoothly quadratic throughout. The `0.08s` came from a benchmark whose 50,000-character input file
**did not exist** — `cat` failed, the string was empty, and the loop timed nothing. I then wrote that
number into the header of the very file whose bug was a false performance claim.

The design conclusion survives (script object: 0.09s / 0.20s / 0.60s / 2.42s at 20k/50k/100k/200k);
the number stated for it did not.

### The old handler was not content-dependent

Measured at 40,960: **21.26s plain against 27.31s at 50% quotes** — flat within noise. The shape that
cost an order of magnitude more on dense input was a *candidate for this rewrite* that sliced runs out
of the full list (0.18s → 3.06s), never anything that shipped. I conflated my own rejected candidate
with the predecessor in four places.

The density assertion stays — it pins the design against that candidate, which is the obvious way to
write this and the tempting one to return to — but it no longer claims to describe history.

### Two mechanism claims, both wrong

- `items a thru b` does **not** cost in proportion to its start offset: 2.16s / 2.20s / 2.30s for
  2,000 slices beginning at 1 / 90,000 / 180,000 of a 200,000 list. It carries a large fixed per-call
  cost instead.
- `count of` on the small buffers is **free**: 0.29s against 0.28s for integer counters. The counters
  are tidiness, not speed.

### One real hazard, fixed

The sentinel was `esc is ""` — a **text** comparison — and `considering`/`ignoring` are dynamically
scoped into called handlers. Under `ignoring punctuation`, both escape strings compare equal to `""`,
and the handler returned:

```
say "hi" b\s
```

Raw quote and raw backslash, from the function whose entire job is escaping them. Now
`(length of esc) = 0`, an integer comparison no attribute reaches. The handler this replaced compared
integers throughout and was immune by construction.

### Tests for the shape this PR introduced, which nothing pinned

Thirteen deterministic flush-boundary cases — cluster straddling a flush, escape at 499/500/501,
astral across the buffer, 600 consecutive escapes crossing the `parts` flush, inputs ending exactly on
a flush — plus the `ignoring` cases. Verified by mutation: dropping the final flush fails 21 of them;
reverting the comparison fails the attribute case.

Also taken: `string id {}` **segfaults osascript** (exit 139, uncatchable by `try`), so the non-zero
guards are load-bearing against a crashed process rather than an error — said where the guard is. The
reserved handler names are now enforced by a test that **runs in CI**, and it is four names, not two
(`resolve-mailbox.applescript` is prepended too). `CLAUDE.md` no longer calls a `-2752` compile error
"silent shadowing" — in a repository whose first principle is loud-over-silent, that was backwards.
Explicit timeouts on the timing tests so only their own assertions can fail.

**Verify:** 534 tests across 28 files, 100% on `src/`, 0 vulnerabilities.

File: tests/bridge/escape-for-json.applescript.test.ts [hunk 2]
```diff
@@ -154,6 +154,138 @@ describe.skipIf(!onMacOS)("escapeForJson, executed", () => {
...
+        expect(escapeAt(run(n, 0x22))).toBe(`${"a".repeat(n)}\\"`);
+      }
+    );
+
+    it.each([498, 499, 500])(
+      "keeps a decomposed accent intact when the flush falls inside it (%i leading)",
+      (n) => {
+        // e + U+0301. At n=499 the cluster straddles the boundary exactly.
+        expect(escapeAt(run(n, 0x65, 0x0301))).toBe(`${"a".repeat(n)}é`);
+      }
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply tests/bridge/escape-for-json.applescript.test.ts:208]: Fixed in 7a8f30e — and this is the file's own rule, which makes it a fair catch rather than a style note. The header states that a literal accented character could be the precomposed U+00E9 or the decomposed pair, that those are exactly the two forms that behave differently here, and that an editor could normalise one into the other. I then put a literal one in an expected value.

Now `\u0065\u0301` on both sides of the assertion. There is a small confirmation that this was not hypothetical: my first patch attempt failed to match the line at all, because the bytes on disk were not the bytes I had typed.

File: tests/bridge/escape-for-json.applescript.test.ts [hunk 3]
```diff
@@ -154,6 +154,138 @@ describe.skipIf(!onMacOS)("escapeForJson, executed", () => {
...
+    );
+
+    it("keeps a ZWJ sequence intact across the boundary", () => {
+      // The emoji occupies 499..502, so the joiner itself crosses the flush.
+      expect(escapeAt(run(498, 0x1f468, 0x200d, 0x1f469))).toBe(
+        `${"a".repeat(498)}\u{1f468}‍\u{1f469}`
+      );
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply tests/bridge/escape-for-json.applescript.test.ts]: Fixed in 7a8f30e. Agreed, and worse than the accent case: an invisible joiner sitting between two astral escapes, in an assertion whose entire point is that the joiner survives a flush boundary. If a formatter dropped it, the test would still pass — against a weaker input than the one it claims to check.

Now `\u200d`. No non-ASCII remains anywhere in that block.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,32 @@ All notable changes to this project will be documented in this file.
...
+
+- Escaping a long message body no longer approaches the bridge's 30s timeout. `escapeForJson` was
+  quadratic in the length of its input: 40,000 characters took ~25s, and a body somewhat larger
+  than that would have failed as a timeout naming neither the tool nor the cause. The same input
+  now escapes in ~0.3s (#42).
+
+  Two quadratic sources had to go together. Element access on a large AppleScript list is O(n) per
+  element — the read loop alone cost 20.46s at 50,000 code points — which the fix avoids by holding
+  the list in a script object property; and unbounded appends grow the accumulator until it re-
+  enters that same trap, which the fix avoids by flushing at a fixed threshold. Element access
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply CHANGELOG.md]: Fixed in 7a8f30e. Correct — the wrap put `re-` at the end of one line and `enters` at the start of the next, which Markdown renders as `re- enters`. Rewrapped so the word stays whole.

File: tests/bridge/prepended-handler-names.test.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * This is the one check in this area that runs **in the merge gate**. Everything else touching
+ * AppleScript needs `osascript` and is skipped on Linux; this reads text, so CI enforces it.
+ */
+const BRIDGE = join(process.cwd(), "src/bridge");
+const DOMAINS = join(process.cwd(), "src/domains");
+
+/** Top-level `on someHandler(` declarations, which are the ones that collide. */
+function handlerNames(source: string): string[] {
+  return [...source.matchAll(/^on\s+([A-Za-z_][A-Za-z0-9_]*)\s*\(/gm)].map((m) => m[1]);
+}
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply tests/bridge/prepended-handler-names.test.ts]: Fixed in 01271ef — the gap is real, and the matcher now catches it. Measured against the prepended `on joinStrings(theList)`:

```
on joinStrings given size:x   -> -2752
on joinStrings above x        -> -2752
```

Neither was detected before: run against a file containing the first, the old matcher returned `[]` and the widened one returns `["joinStrings"]`. Pinned by a mutation check.

One correction worth keeping, because it changes which forms matter. The bare `on <name>` with no parameters is **not** valid AppleScript — it raises `Expected "given", "into", "with", "without" or other parameter name`. The labelled-parameter forms that error message enumerates are the actual gap, and those are what this now covers.

On the `on error` caution: it is anchored to column 1, and measured, no `on error` appears at column 1 anywhere in `src/` — they are all inside handlers and therefore indented. `error` is also excluded by name, so a top-level one could not be read as a declaration either.
